notebooks/working/ho_deconfounder_on_sim_data.py

```diff
@@ -0,0 +1,811 @@
+X_DA = (data[data['mode_id']==1][xs] - data[data['mode_id']==1][xs].mean())/data[data['mode_id']==1][xs].std()
+
+X_SR2 = (data[data['mode_id']==2][xs] - data[data['mode_id']==2][xs].mean())/data[data['mode_id']==2][xs].std()
+
+X_SR3 = (data[data['mode_id']==3][xs] - data[data['mode_id']==3][xs].mean())/data[data['mode_id']==3][xs].std()
+
+# +
+(confounders_DA, holdouts_DA, 
+ holdoutmasks_DA, holdoutrow_DA)= confounder(holdout_portion=0.2, 
+                                       X=X_DA, latent_dim=latent_dim)
+
+(confounders_SR2, holdouts_SR2, 
+ holdoutmasks_SR2, holdoutrow_SR2)= confounder(holdout_portion=0.2, 
+                                       X=X_SR2, latent_dim=latent_dim)
+
+(confounders_SR3, holdouts_SR3, 
+ holdoutmasks_SR3, holdoutrow_SR3)= confounder(holdout_portion=0.2, 
+                                       X=X_SR3, latent_dim=latent_dim)
+
+# +
+data.loc[data['mode_id']==1, 'confounder_da'] = confounders_DA[2]
+data['confounder_da'] = data['confounder_da'].fillna(0)
+
+data.loc[data['mode_id']==2, 'confounder_sr2'] = confounders_SR2[2]
+data['confounder_sr2'] = data['confounder_sr2'].fillna(0)
+
+data.loc[data['mode_id']==3, 'confounder_sr3'] = confounders_SR3[2]
+data['confounder_sr3'] = data['confounder_sr3'].fillna(0)
+
+data['confounder_all_2'] = data[['confounder_da', 'confounder_sr2', 'confounder_sr3']].sum(axis=1)
+data.head()
+
+
+
+
+
+# +
+mnl_specification_causal = mnl_specification_noncausal.copy()
+mnl_names_causal = mnl_names_noncausal.copy()
+
+mnl_specification_causal["confounder_all_2"] = [1, 2, 3]
+mnl_names_causal["confounder_all_2"] = ["Confounder - Drive alone",
+                                     "Confounder - Shared ride 2", 
+                                     "Confounder - Shared ride 3"]
+
+# +
+# Estimate the basic MNL model, using the hessian and newton-conjugate gradient
+mnl_model_causal = cm.create_choice_model(data=data,
+                                   alt_id_col="mode_id",
+                                   obs_id_col="observation_id",
+                                   choice_col="sim_choice",
+                                   specification=mnl_specification_causal,
+                                   model_type="MNL",
+                                   names=mnl_names_causal)
+
+num_vars = len(reduce(lambda x, y: x + y, mnl_names_causal.values()))
+# Note newton-cg used to ensure convergence to a point where gradient 
+# is essentially zero for all dimensions. 
+mnl_model_causal.fit_mle(np.zeros(num_vars),
+                  method="BFGS")
+
+# Look at the estimation results
+mnl_model_causal.get_statsmodels_summary()
+# -
+
+# ## Investigate
+
+# +
+data_da = data[data['mode_id']==1]
+
+data_da.plot.scatter('total_travel_distance', 'confounder_all_2')
+
+# +
+data_mode_specific = data[data['mode_id']==1]
+
+data_mode_specific.plot.scatter('total_travel_distance', 'confounder_all')
+
+
+# -
+
+# ## Putting everything in a function
```

Comment:
Before you conclude the notebook, I think you should have a final cell where you answer "So what?"

In other words, what are the high-level takeaways from using the deconfounder on the simulated data? Did the estimated coefficients have lower bias when using the inferred confounders, relative to simply omitting travel distance? I would include a dataframe that shows:
- the true coefficients used to simulate the data.
- estimated coefficients with all variables (including total travel distance)
- estimated coefficients without total travel distance
- estimated coefficients with inferred confounders using your 1st method
- estimated coefficients with inferred confounders using your 2nd method

Additionally, just for clarity I would make and display a second dataframe that shows the absolute bias in the estimated coefficients, as in take the absolute value of [the first dataframe - the column for the true coefficients used to simulate the data].